### PR TITLE
Performance optimizations

### PR DESCRIPTION
**Pull request by radex**
- Add memory pressure callbacks (they're not actually used yet, but eventually, they'll be hooked up to native machinery on React Native to allow caches to be purged on low memory to keep the app from being killed)
- Make findAndObserve able to emit a value synchronously. By extension, this makes Relations put into withObservables able to render the child component in one shot. Avoiding the extra unnecessary render cycles avoids a lot of DOM and React commit-phase work, which sped up loading some views in NT by 30% !
- Get rid of deep cloning in LokiJS in synchronous mode - we now make invariants about some of this data either being immutable, or we clone only as little as necessary to avoid bugs on mutations. We save ~5ms on load times in some views
- loki encodeQuery is now faster. A little uglier, but still straightforward. This saves 2ms in some views

### CODE REVIEW

File: src/utils/common/memory/index.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+type Callback = () => void
+const lowMemoryCallbacks: Callback[] = []
+
+export function onLowMemory(callback: Callback): void {
+  lowMemoryCallbacks.push(callback)
+}
+
+// TODO: Not currently hooked up to anything
+export function _triggerOnLogMemory(): void {
```
Comment (kokusGr) [inline src/utils/common/memory/index.js:11]:
```suggestion
export function _triggerOnLowMemory(): void {
```

Comment (radex) [reply src/utils/common/memory/index.js:11]: done

File: CHANGELOG-Unreleased.md [hunk 2]
```diff
@@ -12,6 +12,9 @@
 
 ### Changes
 
+- [Performance] findAndObserve is now able to emit a value synchronously. By extension, this makes Relations put into withObservables able to render the child component in one shot. Avoiding the extra unnecessary render cycles avoids a lot of DOM and React commit-phase work, which can speed up loading some views by 30%
+- [Performance] LokiJS is now faster
```
[the remark this answers is not in the document]
Comment (radex) [reply CHANGELOG-Unreleased.md:16]: done

File: src/adapters/lokijs/index.js [hunk 3]
```diff
@@ -89,7 +89,7 @@ export default class LokiJSAdapter implements DatabaseAdapter {
       validateAdapter(this)
     }
 
-    this.workerBridge.send(SETUP, [options], devSetupCallback)
+    this.workerBridge.send(SETUP, [options], devSetupCallback, 'immutable', 'immutable')
```
[the remark this answers is not in the document]
Comment (radex) [reply src/adapters/lokijs/index.js:92]: I agree, but not this PR, it's too big and took too long already. Next time I'll look at this code, I'll make it better

**Comment by kokusGr** [comment]
Thanks, merging.

**Pull request merged**